### PR TITLE
Core separate manager from preview

### PR DESCRIPTION
**Pull request by ndelangen**
I've separated the manager from the preview webpack build.

The manager is now build in a separate child process.

this better isolates the user's code, because the webpack config is no longer shared between manager & preview.

This makes it possible for 2 versions of react to exist: 1 in preview, 1 in manager.
Or the preview react to be swapped with a react compatible lib like preact.

Also, previously conflicting webpack loaders are now possible.
We no longer need to blacklist/whitelist modules in loaders to make sure they do not double apply.

In essence: the users can now be given far more control of the webpack config.

---

This ties into presets.

Preset have the ability to add babel|webpack config to the preview and manager separately.

### CODE REVIEW

File: lib/cli/test/fixtures/react_babel_config_js/package.json [hunk 1]
```diff
@@ -15,6 +15,6 @@
     "@babel/plugin-proposal-optional-chaining": "7.0.0",
     "@babel/core": "7.1.0",
     "babel-loader": "8.0.2",
-    "@storybook/react": "^4.0.0-alpha.17"
+    "@storybook/react": "^4.0.0-rc.6"
```
Comment (Hypnosphi) [inline lib/cli/test/fixtures/react_babel_config_js/package.json]: It's probably my mistake that those dependencies are in fixtures. They should be added by our CLI in test runs instead

**Comment by Hypnosphi** [comment]
> This makes it possible for 2 versions of react to exist: 1 in preview, 1 in manager.

Is this something not implemented here yet?

**Comment by igor-dv** [comment]
> Is this something not implemented here yet?

How can we support #4191 now 🤔 ?

File: app/riot/src/server/framework-preset-riot.js [hunk 2]
```diff
@@ -15,5 +15,12 @@ export function webpack(config) {
         },
       ],
     },
+    resolve: {
+      ...config.resolve,
+      alias: {
+        ...config.resolve.alias,
+        'riot-compiler': 'riot-compiler/dist/es6.compiler',
```
Comment (igor-dv) [inline app/riot/src/server/framework-preset-riot.js]: I am not sure what happened (maybe env changes) but riot-compiler is now resolved to riot-compiler/lib/compiler which is a node version and riot app is failing.
This is change is a workaround, but must be fixed in this PR

File: lib/core/src/server/manager/manager-webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      }),
+      new HtmlWebpackHarddiskPlugin(),
+      new webpack.DefinePlugin({ 'process.env': stringified }),
+      new CaseSensitivePathsPlugin(),
+      new Dotenv({ silent: true }),
+    ],
+    module: {
+      rules: [
+        {
+          test: /\.jsx?$/,
```
Comment (igor-dv) [inline lib/core/src/server/manager/manager-webpack.config.js]: Assuming we have a manager client code already transpiled, we can remove this rule, right?

Comment (ndelangen) [reply lib/core/src/server/manager/manager-webpack.config.js]: Likely, but users could load local un-compiled addons, in addons.js

Comment (igor-dv) [reply lib/core/src/server/manager/manager-webpack.config.js]: 😞

**Comment by ndelangen** [comment]
> > This makes it possible for 2 versions of react to exist: 1 in preview, 1 in manager.
> 
> Is this something not implemented here yet?

It could work in theory already, But we still have to change 'react' to a regular dependency.

File: lib/core/src/server/manager/manager-webpack.config.js (hunk 3 above)
Comment (igor-dv) [reply lib/core/src/server/manager/manager-webpack.config.js]: Should we even handle this usecase ?

Comment (ndelangen) [reply lib/core/src/server/manager/manager-webpack.config.js]: I think we should, it makes it easier to get started writing addons?

File: .babelrc.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+module.exports = {
```
Comment (libetl) [inline .babelrc.js]: should not it be babel.config.js ?

File: app/riot/src/server/framework-preset-riot.js (hunk 2 above)
Comment (libetl) [reply app/riot/src/server/framework-preset-riot.js]: oops, I don't know what I can do about it.

File: .babelrc.js (hunk 4 above)
Comment (ndelangen) [reply .babelrc.js]: Got this from here:
https://babeljs.io/docs/en/config-files

Seems to work..

**Comment by ndelangen** [comment]
@tmeasday @shilman any idea what's wrong with chromatic here?

It seems consistent, but I can't reproduce locally

File: app/riot/src/server/framework-preset-riot.js (hunk 2 above)
Comment (igor-dv) [reply app/riot/src/server/framework-preset-riot.js]: @ndelangen, this issue was not addressed but introduced in this PR...

File: lib/core/src/server/config/utils.js [hunk 5]
```diff
@@ -2,38 +2,37 @@ import path from 'path';
...
-
-export const excludePaths = [path.resolve('node_modules')];
-
 export const nodeModulesPaths = path.resolve('./node_modules');
-
-export const nodePaths = (process.env.NODE_PATH || '')
-  .split(process.platform === 'win32' ? ';' : ':')
-  .filter(Boolean)
-  .map(p => path.resolve('./', p));
+export const excludePaths = [nodeModulesPaths];
```
Comment (igor-dv) [inline lib/core/src/server/config/utils.js]: Can we revert this changes in favor of #4706 ?

File: lib/core/src/server/manager/manager-webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import webpack from 'webpack';
+import Dotenv from 'dotenv-webpack';
+import HtmlWebpackPlugin from 'html-webpack-plugin';
+import HtmlWebpackHarddiskPlugin from '@ndelangen/html-webpack-harddisk-plugin';
+import CaseSensitivePathsPlugin from 'case-sensitive-paths-webpack-plugin';
+
+import { version } from '../../../package.json';
```
Comment (igor-dv) [inline lib/core/src/server/manager/manager-webpack.config.js]: we should provide the version to the child process.

File: lib/core/src/server/config/utils.js (hunk 5 above)
Comment (ndelangen) [reply lib/core/src/server/config/utils.js]: I'm unsure exactly what it'd be that I'd have to change here.

**Comment by igor-dv** [comment]
@ndelangen, I will revert all the NODE_PATH changes today. 
Anyway, there is also this - https://github.com/storybooks/storybook/pull/4590#discussion_r228752008 comment.

**Comment by igor-dv** [comment]
> @ndelangen, I will revert all the NODE_PATH changes today

Actually I think it will be easier to just merge the #4706 before

**Comment by ndelangen** [comment]
Should be good now

**Comment by libetl** [comment]
Congratulations
